course-material/advanced-querying-techniques/subqueries.md

```diff
@@ -1,14 +1,316 @@
+
+
+```{code-cell} ipython3
+%%sql
+CREATE SCHEMA s1;
+CREATE TABLE s1.account AS
+FROM read_csv_auto('expanded_data/account.csv', header=True, sep=',');
+CREATE TABLE s1.district AS
+FROM read_csv_auto('expanded_data/district.csv', header=True, sep=',');
+CREATE TABLE s1.loan AS
+FROM read_csv_auto('expanded_data/loan.csv', header=True, sep=',');
+```
+
+## Queries
+
+Let's first load each table for reference and to better understand their contents.
+
+```{code-cell} ipython3
+%%sql
+SELECT *
+FROM s1.account
+LIMIT 5
+```
+
+```{code-cell} ipython3
+%%sql
+SELECT *
+FROM s1.district
+LIMIT 5
+```
+
+```{code-cell} ipython3
+%%sql
+SELECT *
+FROM s1.loan
+LIMIT 5
+```
+
+### Subquery in SELECT
+
+The query belows results in the average loan amount for each `account_id` by having a subquery in the `SELECT` clause.
+
+```{code-cell} ipython3
+%%sql
+SELECT s1.account.account_id, s1.account.date, 
+      (SELECT AVG(s1.loan.amount) 
+      FROM s1.loan 
+      WHERE s1.loan.account_id = s1.account.account_id
+      ) AS average_loan_amount
+FROM s1.account 
+WHERE average_loan_amount IS NOT NULL
+ORDER BY average_loan_amount
+LIMIT 5
+```
+
+The above SQL statement is quite verbose. Let's instead utilize aliases to make the query more readable.
+
+```{code-cell} ipython3
+%%sql
+SELECT a.account_id,
+      (SELECT AVG(l.amount) 
+      FROM s1.loan AS l 
+      WHERE l.account_id = a.account_id
+      ) AS average_loan_amount
+FROM s1.account AS a
+WHERE average_loan_amount IS NOT NULL
+ORDER BY average_loan_amount
+```
+
+Let's first focus on the "outer query" of this statement. The outer query is everything in the SQL statement besides the second argument in the `SELECT` clause. This outer query results in just the `account_id` from the `s1.account` table. The "inner query" is the second argument in the `SELECT` statement. This query is called for each row from the outer query, which finds the average loan amount for each account by relation in the inner query's `WHERE` clause.
+
+In these datasets, each account actually only has one loan. We solely specify the `AVG()` function for demonstration purposes. We also specify `IS NOT NULL` because only 682 accounts actually have loans. The remaining accounts would have a `NULL` value. 
+
+### Subquery in FROM
+
+In the query below, we find the average loan amount by district name while utilizing a subquery in a `FROM` clause.
+
+```{code-cell} ipython3
+%%sql
+SELECT d.district_name, 
+       ROUND(AVG(inner_query.amount),2) AS average_loan_amount
+FROM 
+    (SELECT l.loan_id, 
+            l.amount, 
+            a.district_id 
+    FROM s1.loan AS l 
+    INNER JOIN s1.account AS a 
+      ON l.account_id = a.account_id
+    ) AS inner_query
+INNER JOIN s1.district AS d
+  ON inner_query.district_id = d.district_id
+GROUP BY d.district_name
+
```

Comment:
Similar comment as above, demonstrate how you can make this more reusable and clean by using the --save option